Create AddMessage function wrapper

Brett208 opened the pull request:
First pass for pushing const_char cast detail to centralized point in OP2Helper. Leverages default function arguments to reduce calling code bulk.

Considered adding 

``` C++
void AddMessage(const char *message, int pixelX, int pixelY, int soundIndex = SoundID::sndMessage2, int toPlayerNum = -1);

void AddMessage(const char *message, LOCATION(), int pixelX, int pixelY, int soundIndex, int toPlayerNum) {
	// message is passed into outpost 2 as a non const argument. The value of message is not modified by Outpost 2.
	TethysGame::AddMessage(pixelX, pixelY, const_cast<char*>(message), toPlayerNum, soundIndex);
}
```

However, I think pixelX and pixelY would need to get passed by a point structure to allow both versions of AddMessage to work properly when using default parameters. I thought about using LOCATION, but it seems to specify tile location, which I didn't want to have the user confused about.

-Brett

Brett208 committed:
Create AddMessage function wrapper

pushes const_char cast detail to centralized point in OP2Helper. Leverages default function arguments to reduce calling code bulk.

Brett208 commented on OP2Helper.h:
```diff
@@ -89,3 +89,10 @@ void RandomizeList(int numItems, ListItemType list[])
+
+// Set toPlayerNum = -1 to send to all players. 
+// Default sound effect is phone ringing (sending text message to other player)
+void AddMessage(const char* message, int soundIndex = SoundID::sndMessage2, int toPlayerNum = -1, Unit sourceUnit = NullUnit);
```
Made toPlayerNum's default -1 instead of 0. I think this makes more sense. In colony games, I suspect -1 will be equivalent to 0. In multiplayer games, the default makes more sense to be all players to me instead of player 0.

Any reason not to use SoundID instead of int for type of soundIndex, and PlayerNum instead of int for type of toPlayerNum?

Thanks,
Brett

DanRStevens replied:
I like changing the default to `-1` / `PlayerAll`. Let's do that.

Probably also a good idea to be more explicit about types. I like the idea of using `SoundID` and `PlayerNum`. They're regular enums, so they should convert between int. That means the use of enums shouldn't get in the way if someone really wants to use ints somewhere. Might want to test that though.

DanRStevens commented on OP2Helper.h:
```diff
@@ -89,3 +89,10 @@ void RandomizeList(int numItems, ListItemType list[])
 void CenterViewOnPlayerCC();
+
+// Use when passing a non-existing unit.
+const Unit NullUnit;
```
Will this guarantee zero initialization like this?

DanRStevens replied:
Unit has a default constructor, which does zero initialized.

DanRStevens commented:
We can add variants for tile or pixel locations. We should be consistent about parameter order though. I don't want to insert different parameters in the middle of the parameter list.

It could be that `LOCATION` is for tiles, which get converted to pixels (`tileX/Y * 32 + 16`). Name the parameter to be explicit if it is in tiles or in pixels.

DanRStevens commented:
I noticed this isn't specified in the SDK headers, but a `pixelY` value of `-1` will add the message without associating with any particular point on the map. It disables the goto feature when you click (or double click?) on the message, (or hit spacebar). It may also affect the volume of the sound. Sounds associated with a certain location are louder when you are viewing that location. General game sounds without a location are always full volume.

Typically, levels or internal code will pass (-1, -1) as the pixel location for general map messages.

I'm not sure if `NullUnit` has coordinates set to (-1, -1) or not. If not, perhaps we should translate `NullUnit` parameters into a call with coordinates (-1, -1).

Brett208 commented:
Thanks for researching the AddMessage function. I pushed a PR in Outpost2DLL adding a comment about setting the location to -1. I don't understand how to easily set a Unit to position -1, -1. Sounds worth updating to do correctly though?

Brett208 commented:
I don't think setting LOCATION directly in pixels would be intuitive. If we stick to tiles for LOCATION, the formula of `tileX/Y * 32 + 16` it would not produce (-1, -1) when a tile LOCATION of (-1, -1) was specified. I think we want to support the location agnostic message.

I guess we add a check inside the function to see if a tile location of (-1, -1) is passed and manually address. Feels a bit like a hack. Maybe there is a better way?

pseudocode
```C++
if (loc == LOCATION(-1,-1) {
    pixelX = -1;
    pixelY = -1;
}
else {
    pixelX = loc.X * 32 + 16;
    pixelY = loc.Y * 32 + 16;
}
```

-Brett

DanRStevens commented:
Ahh, I meant as an `if` check for `NullUnit`, much like you did for `LOCATION`:
```cpp
if (unit == NullUnit) {
  AddMessage(-1, -1, ...);
} else {
  AddMessage(unit, ...);
}
```
Not sure if it's needed though. Maybe they were clever enough to set the location of Unit 0 to (-1, -1) already.

----

Could do similar for `LOCATION`, with two calls, and inline expansion of the coordinate values. Though in that case, they're both calling the same version of the method from Outpost2.exe. As such, I can see why you'd want to adjust the coordinates up front, and then make one call.

I don't see a particularly better way to handle the LOCATION tile to pixel conversion.

----

Maybe we could define a `LOCATION NullLocation(-1, -1);` object.

Brett208 committed:
Add location based AddMessage wrapper function

Brett208 committed:
Improve function arguments on AddMessage wrappers

Brett208 committed:
Add global LOCATION == and != operators

Brett208 commented:
Thought about this some more. I think we want NullLocation and not NullUnit. If someone wants to pass a location/unit agnostic message, they should use the AddMessage function that doesn't include a unit.

I think it is better if we use int instead of SoundID and PlayerNum. Otherwise, someone has to cast the playerIndex to a PlayerNum when passing, which is okay. But it is also natural to pass PlayerNum as an int I think?

I defined a LOCATION == and != operator. If these don't look right, no problem deleting the last commit as it is outside scope of this work, just seemed useful.

-Brett

DanRStevens commented on OP2Helper.cpp:
```diff
@@ -110,6 +110,15 @@ LOCATION operator- (const LOCATION &loc1, const LOCATION &loc2)
+
+bool operator!= (const LOCATION& loc1, const LOCATION &loc2)
+{
+	return (loc1.x != loc2.x || loc1.y != loc2.y);
```
This could just delegate and use:
```cpp
return !(loc1 == loc2);
```
Less code to maintain.

Brett208 replied:
I'll make the change shortly

DanRStevens reviewed (commented):
I'm liking how this is shaping up.

----

I think you're right about scrapping `NullUnit`.

----

I think you're right about there being some problems with `enum` parameters. They will auto convert to `int`, but they won't auto convert from `int`. Maybe it can be worked around with a type conversion operator overload. Not sure if we want to go there or not. The code might look something like:
```cpp
EnumType operator EnumType(int value) {
  return static_cast<EnumType>(value); // Optional: throw for out of range values
}
```
I'm uncertain if that code needs an explicit type cast to be called, or if it can be implicitly called. Either way I'm fine leaving it out.

At any rate, this needs to work for `int`, without burdening the user with an explicit type cast. I'd rather just declare the parameter as `int` than burden the user.

----

The `LOCATION` operators are a good idea. I can see that being generally useful.

----

Perhaps we should consider an `AddMessage` overload that doesn't take a position, and auto fills in values of (-1, -1). That might be a way of avoiding the `if` code block. Probably better practice to call a different method, than to overload a parameter value with a special sentinel meaning. To avoid parameter ambiguity, we could have a slightly different name for the two versions:

```cpp
// Message tied to a map location (supports GoTo, volume influenced by current view location)
void AddMapMessage(const char* message, Unit sourceUnit, int soundIndex = SoundID::sndMessage2, int toPlayerNum = PlayerNum::PlayerAll);
void AddMapMessage(const char* message, const LOCATION& location, int soundIndex = SoundID::sndMessage2, int toPlayerNum = PlayerNum::PlayerAll);
void AddMapMessage(const char* message, int pixelX, int pixelY, int soundIndex = SoundID::sndMessage2, int toPlayerNum = PlayerNum::PlayerAll);
// General game message, not tied to a map location (always full volume)
void AddGameMessage(const char* message, int soundIndex = SoundID::sndMessage2, int toPlayerNum = PlayerNum::PlayerAll);
```

If we go this route, maybe we should scrap `NullLocation`? I don't think that particular value is used elsewhere.

Brett208 commented:
I like the function: 

```C++
void AddGameMessage(const char* message, int soundIndex = SoundID::sndMessage2, int toPlayerNum = -1);

```

And I like that it eliminates NullLocation.

I cannot think of an instance where I would be working in pixels instead of location on a map. @DanRStevens, do you see it as useful? Maybe we can leave the following function out for now. Then again it isn't too big of a deal to add. as it is a single line of code inside the definition.

```C++
void AddMapMessage(const char* message, int pixelX, int pixelY, int soundIndex = SoundID::sndMessage2, int toPlayerNum = -1);
```

Brett208 committed:
Reduce amount of code in LOCATION global operator overloads

DanRStevens commented:
The game internally tracks things in pixels. If you ever want to associate a message based on an internal pixel value, it would be easier if that method was available. Then you wouldn't need to convert back to tiles, just to have things rounded and then converted back to pixels.

Mostly though, it feels somehow incomplete if we dropped it.

And yeah, it's super easy to add.

----

Just realized my example code wasn't using `PlayerNum::PlayerAll`.

Brett208 committed:
Merge branch 'master' into CreateAddMessageWrapper

Brett208 committed:
Remove appveyor file

Brett208 committed:
Add more variations to AddMessage function series

DanRStevens commented on OP2Helper.cpp:
```diff
@@ -129,20 +129,30 @@ void CenterViewOnPlayerCC() {
+	if (location != LOCATION(-1,-1)) {
 		pixelX = location.x * 32 + 16;
 		pixelY = location.y * 32 + 16;
 	}
```
Should we remove this if block and do the calculation unconditionally? If they really didn't want to be specifying a location, they would probably use the other overload. I'm thinking we should simplify.

DanRStevens commented on .appveyor.yml:
```diff
@@ -1,6 +0,0 @@
-  - Release
-  - Debug
-build:
-  project: OP2Helper.sln
```
After merging master, and then removing the `.appveyor.yml` file on this branch, this would cause cause `.appveyor.yml` to be removed from master when this branch is merged to master. I assume that wasn't quite the intention here.

As for getting AppVeyor builds working, the project still needs to be added on the AppVeyor website. Adding the project there installs any needed hooks on GitHub, so AppVeyor will be informed when a new commit is pushed. I just added the project over on AppVeyor.

DanRStevens committed:
Revert "Remove appveyor file"

This reverts commit 0ebff64f6ea7a771aebee3d34cb411e860a65f13.

DanRStevens committed:
Simply AddMapMessage by removing if block

If the caller does not intend to set a specific location, they could
have called `AddGameMessage` instead.

DanRStevens committed:
Update cast comment

DanRStevens committed:
Update AppVeyor config to clone Outpost2DLL

DanRStevens commented:
Ahh, interesting problem I ran into with AppVeyor. Thankfully not too hard to find a solution. I assume that's what you'd originally intended.

I went ahead and made the change for my other suggestion.

If you agree, feel free to go ahead and merge.
